Add file checksum to de-dup file stores

ligangty opened the pull request:
* Add new filechecksum table, stores (checksum, fileid, storage)
  * When write new file, will check checksum against checksum table for
duplication verification
  * If checksum dupe, will use existed fileid and storage for new file
as file path.
  * When deletion, only when all fileids are not referenced, then the checksum
and storage can be removed.

ligangty committed:
Add file checksum to de-dup file stores

  * Add new filechecksum table, stores (checksum, fileid, storage)
  * When write new file, will check checksum against checksum table for
duplication verification
  * If checksum dupe, will use existed fileid and storage for new file
as file path.
  * When deletion, only when all fileids are not referenced, then the checksum
and storage can be removed.

ruhan1 commented on src/main/java/org/commonjava/storage/pathmapped/datastax/CassandraPathDB.java:
```diff
@@ -73,21 +86,29 @@ public CassandraPathDB( PathMappedStorageConfig config )
 
-        preparedListQuery = session.prepare(
-                        "SELECT * FROM " + keyspace + ".pathmap WHERE filesystem=? and parentpath=?;" );
+        preparedFileChecksumByChecksum = session.prepare( "SELECT * FROM " + keyspace + ".filechecksum WHERE checksum=? ALLOW FILTERING;" );
```
"ALLOW FILTERING" is not needed because checksum is the primary key.

ligangty replied:
Actually checksum is not. The fileid is because it is unique in filecheckum table.

ligangty replied:
And I'm not sure now how to solve this "ALLOW FILTERING" problem. Maybe we need another reverse table here to store the first fileid with the checksum as primary key, and use the filechecksum table to just store the checksum with all matched fileids. And maybe the storage path should be stored in the reverse table.

ruhan1 replied:
I think the checksum should be the primary key. The checksum is unique too. We just need to get the fileid/storage by checksum.

ruhan1 replied:
If MD5 is not that sure, we can use sha1 or sha256. Basically, we can simply ignore the possibility of sha collision.

ligangty replied:
No, it's not related to algorithm, it's about the data itself. In this table, the checksum to fileid is one-to-many, that means there will be more than one fileids mapping to same checksum. So the records should be like:
checksum1-fileid1
checksum1-fileid2
checksum1-fileid3
....
So the checksum is obviously not unique here.

ruhan1 replied:
I don't understand why that happens. Can you give an example?

ruhan1 replied:
IMO, if we find the checksum is the same, we should keep using fileid1.

ligangty replied:
For example, if we store these three files:
a/b/samefile.txt, c/d/samefile.txt, e/f/samefile.txt
These three files are same but with different path, so their fileids are not same (let's use fileid1, fileid2, fileid3), but checksum is same(let's use "md5-checksum" for example). So when store these in filechecksum table, the recorded data will be:
md5-checksum : fileid1
md5-checksum : fileid2
md5-checksum : fileid3
That's what I'm saying in the example. And obviously the checksum is not unique here.
So I'm thinking we should add a table here to mapping checksum to store path, in this table the checksum will be primary key.

ruhan1 replied:
Aha, that is exactly how we use this de-dup. 
These three files are same but with different path, so their fileids are **same** :) :) :) 
e.g, when store a/b/samefile.txt, we get id1 and checksum xyz. When store others, we use xyz to check and find id1, then we use id1. With same files, we use id1 for ever.

ligangty replied:
Ok, that means the fileid is physical file bounded but not abstract path bounded. Maybe I have something wrong for understanding this fileid. I need to think more about it.

ruhan1 replied:
Yep. One fileid is equal to one physical file. 
Multiple paths can map to single fileId (what we do in copy()).

ruhan1 commented on src/main/java/org/commonjava/storage/pathmapped/datastax/CassandraPathDB.java:
```diff
@@ -206,13 +227,31 @@ public void insert( PathMap pathMap )
+        if ( existedFileChecksum != null )
+        {
+            logger.info( "File checksum conflict, should use existed file storage" );
+            final PathMap existed = getPathMapByFileId( existedFileChecksum.getFileId() );
```
I can say there are two issues. The mapping to a fileId is many-to-one, i.e., there might be multiple entries in pathMap table that map to one fileId. The other issue is "previousStorage = existed.getFileStorage()". It is set in following "setFileStorage" but reclaimed at the end. I guess it should be "reclaim(deprecatedFileId, newStorage);"?
Anyway, this is easy to correct. The main question is how to get the existing file storage. We can add a field "storage" to Checksum table. So each time we get by checksum, we get both the fileId and storage. This also eliminate the ALLOW FILTERING problem.

ligangty replied:
For first one, seems it is a typo..... I'll fix it.
For second one, seems it is good idea to add a storage field to avoid filtering. I'll add it here.

ruhan1 commented on src/main/java/org/commonjava/storage/pathmapped/datastax/CassandraPathDB.java:
```diff
@@ -73,21 +86,29 @@ public CassandraPathDB( PathMappedStorageConfig config )
+        preparedListQuery =
+                session.prepare( "SELECT * FROM " + keyspace + ".pathmap WHERE filesystem=? and parentpath=?;" );
+
+        preparedPathMapByIdQuery =
```
Aha, this is what we should avoid. The "ALLOW FILTERING" could be a big performance killer. In Cassandra, this only should be used when "you know and can live with the performance penalty". In other words, if a table is small-sized and the cluster nodes are less and so on, use it after testing the performance. Basically, we should only query based on the "cluster key" fields (the fields in the PRIMARY key following the partition key). We can solve this problem in another way. I will comment it below on the insert code.

ligangty replied:
Ok, that means we need to do redundancy for the field we need to do filtering.  Maybe just as said as above, a storagepath field is needed in separate table.

ruhan1 commented on src/main/java/org/commonjava/storage/pathmapped/util/PathMapUtils.java:
```diff
@@ -14,6 +14,16 @@
+     * with a successful entry in the de-dupe process firstly but used a existed file storage path, we need to create a new
+     * entry point to that deprecated file storage path created and let gc clean it later. So add this filesystem for this purpose.
+     */
+    public final static String DEDUPE_DEPRECATE_FILESYSTEM = "DEDUPE_CHECKSUM_FILESYSTEM";
```
this seems not used?

ligangty replied:
Yup. In old thought I'm thinking if a new pathmap of deprecated needed with this, but after checking we just need to do a reclaim but not a deprecated pathmap. Forgot to remove this.

ligangty commented:
Updated new code:
 * Added storage in filechecksum table for quick reference.
 * Added index on filechecksum(fileid) to avoid filtering, which means we still need to refer checksum by fileid(in deletion)
 * New u-test to check the correctness of the deletion process with file-checksum(to ensure not to delete file by mistake)

ruhan1 reviewed (approved):
LGTM.
btw, preparedChecksumByFileidQuery not used anymore. please remove it. Feel free to merge it then.

ligangty commented:
The preparedChecksumByFileidQuery is used in deletion. And an index is added to void "ALLOW Filtering" for it.

ruhan1 commented:
Okay. This needs a bit refine then. Cassandra indexing is not as good as database. Just don't trust it, it has same performance issue as ALLOW FILTERING. :) 
We can just insert one more field "checksum" into pathMap and use it in delete.

ligangty commented:
Ok. In updated version I've added checksum to pathmap table to remove the last non-primary-key query on filechecksum table in deletion process.

ruhan1 reviewed (approved):
LGTM!
